### PR TITLE
cql3: Ignore regular index in multi-column WHERE

### LINKED ISSUE #9085: Multi column restrictions ignored on indexed queries
Multi column restrictions seem to be ignored on queries that use an index:

```cql
DROP KEYSPACE IF EXISTS ks;
CREATE KEYSPACE IF NOT EXISTS ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};
CREATE TABLE IF NOT EXISTS ks.t (p int, c1 int, c2 int, c3 int, r1 int, r2 int, primary key (p, c1, c2, c3));
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 0, 0, 0, 0, 0);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 1, 0, 0, 0, 0);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 2, 2, 0, 0, 0);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 0, 1, 0, 0, 1);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 1, 1, 0, 0, 1);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 3, 3, 0, 0, 1);

-- Uncomment for a global index:
CREATE INDEX IF NOT EXISTS ON ks.t(r2);

-- Uncomment for a local index:
-- CREATE INDEX IF NOT EXISTS ON ks.t((p), r2);
SELECT * FROM ks.t WHERE p = 0 and (c1, c2) < (1, 0) and r2 = 0;
```
Prints out:
```cql
 p | c1 | c2 | c3 | r1 | r2
---+----+----+----+----+----
 0 |  0 |  0 |  0 |  0 |  0
 0 |  1 |  0 |  0 |  0 |  0
 0 |  2 |  2 |  0 |  0 |  0
```
(Sometimes the index doesn't build in time and the query returns empty rows. In that case just perform the `SELECT` one more time)

This is wrong - `(1, 0)` and `(2, 2)` are not less than `(1, 0)`

### Versions
* `master` (aaf35b5ac273afd01a73c93c09f27179bdea17b5) - broken
* `Docker Scylla version 4.4.3-0.20210609.bfafb8456` - broken
* `Docker Scylla version 4.3.5-0.20210614.690a96ff5` - works, but requires `ALLOW FILTERING`

### PR DESCRIPTION
**Pull request by dekimir**
When the WHERE clause contains a multi-column restriction as well as
an equality on an indexed regular column, we will not use the index.
The index cannot eliminate post-fetch filtering here, but it does add
an extra table query to the execution.

This is similar to db63b4034, which dealt with a clustering-column
index.  Here we deal with a non-key column being indexed.

Fixes #9085.

Tests: unit (dev, debug)

Signed-off-by: Dejan Mircevski <dejan@scylladb.com>

### CODE REVIEW

File: test/cql-pytest/test_secondary_index.py [hunk 1]
```diff
@@ -241,3 +241,15 @@ def test_filter_cluster_key(cql, test_keyspace):
...
+def test_multi_column(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, 'p int, c1 int, c2 int, r int, primary key(p,c1,c2)') as tbl:
+        cql.execute(f'CREATE INDEX ON {tbl}(r)')
+        cql.execute(f'INSERT INTO {tbl}(p, c1, c2, r) VALUES (1, 1, 1, 0)')
+        cql.execute(f'INSERT INTO {tbl}(p, c1, c2, r) VALUES (1, 1, 2, 1)')
+        cql.execute(f'INSERT INTO {tbl}(p, c1, c2, r) VALUES (1, 2, 1, 0)')
+        with pytest.raises(InvalidRequest, match="use ALLOW FILTERING"):
+            cql.execute(f'SELECT c1 FROM {tbl} WHERE p=1 AND (c1,c2)<(2,0) AND r=0')
+        # Uncomment when #6200 is fixed:
+        #assert_rows(cql.execute(f'SELECT c1 FROM {tbl} WHERE p=1 AND (c1,c2)<(2,0) AND r=0 ALLOW FILTERING'), [1])
```
Comment (psarna) [inline test/cql-pytest/test_secondary_index.py:255]: It would be slightly better to split this test into two cases - one for the invalid request, which passes, and one for the case which would only work after 6200 gets fixed, marked with `pytest.mark.xfail`. That way it will be easy to spot that the test starts passing, otherwise this commented line will probably stay in here forever, since nobody's going to remember its existence.

It's a nonblocking nitpick though, so I'll approve this series as a whole.

**Review by psarna — APPROVED** [review]

File: test/cql-pytest/test_secondary_index.py (hunk 1 above)
Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py]: Done.

**Comment by dekimir** [comment]
v2:
42ff79e2a * Split testcase into two (valid and xfail)

File: test/cql-pytest/test_secondary_index.py [hunk 2]
```diff
@@ -241,3 +241,23 @@ def test_filter_cluster_key(cql, test_keyspace):
         stmt = SimpleStatement(f"SELECT c1, c2 FROM {table} WHERE c1 = 1 and c2 = 1 ALLOW FILTERING")
         rows = cql.execute(stmt)
         assert_rows(rows, [1, 1])
+
+# Reproduces #9085.
+def test_multi_column_skips_regular_index(cql, test_keyspace):
```
Comment (psarna) [inline test/cql-pytest/test_secondary_index.py:246]: I tested this case against Cassandra (4.0-rc2) and it failed - perhaps you need to add the `scylla_only` fixture?

Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py]: Done.

**Comment by dekimir** [comment]
v3:
271584a00 * Make the test run only on Scylla

File: cql3/restrictions/statement_restrictions.cc [hunk 3]
```diff
@@ -372,6 +372,7 @@ statement_restrictions::statement_restrictions(database& db,
     const expr::allow_local_index allow_local(
             !_partition_key_restrictions->has_unrestricted_components(*_schema)
             && _partition_key_restrictions->is_all_eq());
+    _has_multi_column = find_atom(_clustering_columns_restrictions->expression, expr::is_multi_column);
```
Comment (nyh) [inline cql3/restrictions/statement_restrictions.cc:375]: I can't comment on the commit message so I'll do it here:
Reading the commit message, it seems like this is some sort of optimization of whether or not to use the index. But the linked issue suggests this is actually a correctness bug! So the commit message should have mentioned the error, not just the optimization angle.

File: test/cql-pytest/test_secondary_index.py [hunk 4]
```diff
@@ -241,3 +241,23 @@ def test_filter_cluster_key(cql, test_keyspace):
         stmt = SimpleStatement(f"SELECT c1, c2 FROM {table} WHERE c1 = 1 and c2 = 1 ALLOW FILTERING")
         rows = cql.execute(stmt)
         assert_rows(rows, [1, 1])
+
+# Reproduces #9085.  Scylla-only because Cassandra uses indices differently.
```
Comment (nyh) [inline test/cql-pytest/test_secondary_index.py:245]: I don't understand this test. The example in 9085 shows wrong output. Why don't you just test that? It seems you are just testing that ALLOW FILTERING is needed. That's perhaps a good test for test_allow_filtering.py, but here it would be better to add ALLOW FILTERING and test that the **result** is correct - and should be correct also in Cassandra. Making a correctness test scylla_only doesn't make sense.

File: test/cql-pytest/test_secondary_index.py [hunk 5]
```diff
@@ -241,3 +241,23 @@ def test_filter_cluster_key(cql, test_keyspace):
...
+def test_multi_column_skips_regular_index(cql, test_keyspace, scylla_only):
+    with new_test_table(cql, test_keyspace, 'p int, c1 int, c2 int, r int, primary key(p,c1,c2)') as tbl:
+        cql.execute(f'CREATE INDEX ON {tbl}(r)')
+        cql.execute(f'INSERT INTO {tbl}(p, c1, c2, r) VALUES (1, 1, 1, 0)')
+        cql.execute(f'INSERT INTO {tbl}(p, c1, c2, r) VALUES (1, 1, 2, 1)')
+        cql.execute(f'INSERT INTO {tbl}(p, c1, c2, r) VALUES (1, 2, 1, 0)')
+        with pytest.raises(InvalidRequest, match="use ALLOW FILTERING"):
+            cql.execute(f'SELECT c1 FROM {tbl} WHERE p=1 AND (c1,c2)<(2,0) AND r=0')
+
+@pytest.mark.xfail(reason="Other restrictions are ignored when clustering key is constrained.  Issue #6200")
```
Comment (nyh) [inline test/cql-pytest/test_secondary_index.py:255]: This is a test for a **different** issue, right (this is what you said in #9085 - that #6200 is different).
So please put this test in a separate commit, with a separate commit message and adding "Refs #6200" - so we'll remember that you added a test for 6200 when we come to work on it.

Comment (nyh) [reply test/cql-pytest/test_secondary_index.py:255]: Also, please add a comment in front (or if you must, a docstring) for the test explaining what it tests. At the very minimum, add a statement that it reproduces issue #6200. Yes, you said this in the xfail message, but the xfail will eventually be removed (that's our goal :-)), but the comment explaining why this test exists should stay behind.

**Review by nyh — CHANGES_REQUESTED** [review]

File: test/cql-pytest/test_secondary_index.py (hunk 4 above)
Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py:245]: > The example in 9085 shows wrong output. Why don't you just test that?

I do, below.  But the root cause of #9085 is using the index when we shouldn't.  The fix, then, is to ignore the index.  Which, unfortunately, lands us exactly in the #6200 scenario. :(  We cannot obtain correct results until we fix #6200.  We can, however, ensure that the index isn't used.  I'm happy to move it to test_allow_filtering.py.

File: test/cql-pytest/test_secondary_index.py (hunk 5 above)
Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py:255]: > This is a test for a **different** issue, right (this is what you said in #9085 - that #6200 is different).

No, this is exactly the test you asked about before.  It was all a part of the same test until @psarna [suggested](https://github.com/scylladb/scylla/pull/9115#discussion_r679946627) I split them.

> Also, please add a comment in front (or if you must, a docstring) for the test explaining what it tests.

What would the comment say that isn't already in the test's name?

Comment (nyh) [reply test/cql-pytest/test_secondary_index.py:255]: You wrote in the commit message "Fixes #9085". Now I understand it fixes your new definition of 9085, but it does not fix the specific example which started 9085 - which is still broken and now you call it 6200. If that's the way you decided to go, please fix the issues to follow your decision: Copy the broken example to issue 6200, close 9085 as a duplicate of 6200, and open a new issue with what you are fixing here. Alternatively, I can accept that you don't bother with fixing the issues if at least you fix the code comments and commit messages to explain what's going on.
2. 

 As a reviewer (or somebody who will read this code, or commit, or issue a year from now) I have no idea what's the difference between 9085 or 6200 or which one this commit fix - unless you explain it clearly. I understand

Comment (nyh) [reply test/cql-pytest/test_secondary_index.py:255]: > > Also, please add a comment in front (or if you must, a docstring) for the test explaining what it tests.
> 
> What would the comment say that isn't already in the test's name?

First, I answered exactly that question in the next sentence in my review: *"At the very minimum, add a statement that it reproduces issue #6200. Yes, you said this in the xfail message, but the xfail will eventually be removed (that's our goal :-)), but the comment explaining why this test exists should stay behind."*

Second, are you seriously trying to suggest that from the name "test_multi_column_skips_regular_index" a year from now anybody will understand why this test exists? That it's related to issues 6200 or 9085? Understand why a "multi_column" (multi column what?) should (or shouldn't?) skip a regular index (what's a regular index?), and what happens if it does the wrong thing?
The irony is that you did write a comment of sorts in the xfail message: "Other restrictions are ignored when clustering key is constrained. Issue #6200". I think it's still too short for my taste but it's much better than nothing. The problem with xfail message is that it will be deleted when the bug is fixed - and I want to have a comment that will stay behind. We should know that a test was written for issue 6200 long after this issue is closed - this can help us in the future if we need to backport a bugfix, or if a test starts to fail again and we want to remember why this test even exists and whether the test is even testing a correct thing and why.

File: test/cql-pytest/test_secondary_index.py (hunk 4 above)
Comment (nyh) [reply test/cql-pytest/test_secondary_index.py:245]: I don't know if it should move to test_allow_filtering.py. That file is just for testing whether or not scylla requires "ALLOW FILTERING" on specific queries - it's not about internal implementation details but really visible ALLOW-FILTERING-requirement differences between Scylla and Cassandra. Is this the case here?

But more importantly, I still don't understand why this test is "scylla-only". Is it because we consider Scylla's behavior correct and Cassandra's behavior incorrect? If this is the case, we should mark it cassandra-bug, not scylla-only (but should only do this after careful thinking). Or, alternatively, maybe you consider Scylla's behavior is wrong here but you want to keep it **for now**? In that case, the test should pass on Cassandra's behavior - not Scylla's behavior - and should xfail on Scylla. You can see a few tests like that in `test_allow_filtering.py`.

So in both cases I described using `scylla_only` for such test is wrong. `scylla_only` makes sense mostly for testing Scylla-only features, and sometimes also for cases where we deliberately decided to diverge from Cassandra - not temporarily, not until we fix the code - but a deliberate decision that Cassandra did something badly and we can do it better and will never want to go back to their behavior. Is this the case here?

File: test/cql-pytest/test_secondary_index.py (hunk 5 above)
Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py:255]: > > What would the comment say that isn't already in the test's name?
> 
> First, I answered exactly that question in the next sentence in my review: _"At the very minimum, add a statement that it reproduces issue #6200.

That statement would be wrong.  This reproduces #9085.  #6200 is just a confounding factor here.

> Second, are you seriously trying to suggest that from the name "test_multi_column_skips_regular_index"

Um, that's... not the name we're discussing.

> Understand why a "multi_column" (multi column what?)

Multi-column is only used in one sense in all source and documentation I've ever seen.

File: test/cql-pytest/test_secondary_index.py (hunk 4 above)
Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py:245]: > But more importantly, I still don't understand why this test is "scylla-only". Is it because we consider Scylla's behavior correct and Cassandra's behavior incorrect? 

We definitely consider Scylla's behaviour correct.  And if Cassandra's indices were identical to Scylla's indices, we'd consider Cassandra's behaviour incorrect.  But I don't know enough about Cassandra's indices; for all I know, they may be sufficiently powerful to eliminate the filtering step here.

**Comment by dekimir** [comment]
Please ignore this for a bit.  I just developed a better fix that I'm currently testing and will submit separately.

**Comment by dekimir** [comment]
I'm retracting this approach in favour of #9122.

File: test/cql-pytest/test_secondary_index.py (hunk 5 above)
Comment (dekimir) [reply test/cql-pytest/test_secondary_index.py:255]: > You wrote in the commit message "Fixes #9085". Now I understand it fixes your new definition of 9085, but it does not fix the specific example which started 9085

While I disagree with the rest of your analysis, this tidbit is absolutely valid.  Thinking about how to address it, I figured out a much better fix.  Thank you. :)
